test/common/thread_local/thread_local_impl_test.cc

```diff
@@ -82,6 +82,42 @@ TEST_F(ThreadLocalInstanceImplTest, All) {
   tls_.shutdownThread();
 }
 
+// Validate ThreadLocal::runOnAllThreads behavior with all_thread_complete call back.
+TEST_F(ThreadLocalInstanceImplTest, RunOnAllThreads) {
+  SlotPtr tlsptr = tls_.allocateSlot();
+
+  EXPECT_CALL(thread_dispatcher_, post(_));
+  EXPECT_CALL(main_dispatcher_, post(_));
+
+  // Ensure that the thread local call back and all_thread_complete call back are called.
+  struct {
+    std::atomic<uint64_t> thread_local_calls_{0};
+    std::condition_variable condvar_;
+    std::mutex condvar_mutex_;
+    bool all_threads_complete_;
+  } thread_local_data;
+
+  tlsptr->runOnAllThreads(
```

Comment:
currently it is 1. The thing that is being verified here is whether the main call back is called after the all post callbacks are executed and it is called as per the expected number of times asserted via thread_local_calls_. Both the things are happening. 
The threads here are MockDispatchers so I am not really sure having 10 threads helps. Let @mattklein123  review this test and if he suggests it helps to add 10 threads, I will surely add tomorrow.

Reply:
OK.  If there is only one thread in the system I'm pretty convinced we need no condvars or atomics.

But since this is a test of threading infrastructure, I would hope we'd use real threads for this.

Reply:
@mattklein123 I implemented similar to what you have described. The unit testing environment, the post to main_dispatcher is not actually posting with DispatcherImpl as expected. I tried to post on Main Dispatcher on Dispatcher test also but it did not work. So I added two tests one which actually verifies the call back sequence is correct and another test which verifies the updates with multiple threads. I think it should be good enough. Let me know if you see I am doing some thing wrong in the test.